openforcefield/utils/toolkits.py

```diff
@@ -3054,6 +3100,52 @@ def to_inchikey(self, molecule, fixed_hydrogens=False):
             inchi_key = Chem.MolToInchiKey(rdmol)
         return inchi_key
 
+    def get_tagged_smarts_connectivity(self, smarts):
+        """
+        Returns a tuple of tuples indicating connectivity between tagged atoms in a SMARTS string. Does not
+        return bond order.
+
+        Parameters
+        ----------
+        smarts : str
+            The tagged SMARTS to analyze
+
+        Returns
+        -------
+        unique_tags : tuple of int
+            A sorted tuple of all unique tagged atom map indices.
+        tagged_atom_connectivity : tuple of tuples of int, shape n_tagged_bonds x 2
+            A tuple of tuples, where each inner tuple is a pair of tagged atoms (tag_idx_1, tag_idx_2) which are
+            bonded. The inner tuples are ordered smallest-to-largest, and the tuple of tuples is ordered
+            lexically. So the return value for an improper torsion would be ((1, 2), (2, 3), (2, 4)).
+
+        Raises
+        ------
+        SMIRKSParsingError
+            If RDKit was unable to parse the provided smirks/tagged smarts
+        """
+        from rdkit import Chem
+        from openforcefield.typing.chemistry import SMIRKSParsingError
+        ss = Chem.MolFromSmarts(smarts)
+
+        if ss is None:
+            raise SMIRKSParsingError(f"RDKit was unable to parse SMIRKS {smarts}")
+
+        unique_tags = set()
+        connections = set()
+        for at1 in ss.GetAtoms():
+            if at1.GetAtomMapNum() == 0:
```

Comment:
Are lines like these meant to simply ignore cases where an atom has `:0` as its tag, since that's not part of our spec?

Reply:
Yes. IIRC, RDKit assigns map values of 0 to 
* atoms that don't have map indices at all, and
* atoms like `[#6:0]`
so we play it safe and assume that atoms with map index==0 aren't supposed to be tagged at all.

FYI: Atom map indices are a bit fuzzy right now, because they don't have chemical meaning, but different toolkits have different behaviors around them. So one place we deal with them is in SMIRKS (the topic here), but there's an even bigger question about how to handle them in molecules. I think we know enough about _our_ use cases to lay down a map index spec in OFFMols that will work for the rest of time, but we need to finish the discussion about intended behavior in #522 before we implement it.